distributed/protocol/utils.py

```diff
@@ -81,3 +83,108 @@ def unpack_frames(b):
+    """
+    if not mvs:
+        return memoryview(bytearray(0))
+    if len(mvs) == 1:
+        return mvs[0]
+
+    first = mvs[0]
+    obj = first.obj
+    format = first.format
+
+    first_start_addr = 0
+    nbytes = 0
+    for i, mv in enumerate(mvs):
+        if mv.nbytes == 0:
+            continue
+
+        if mv.obj is not obj:
+            raise ValueError(f"{i}: memoryview has different buffer: {mv.obj} vs {obj}")
+        if not mv.contiguous:
+            raise ValueError(f"{i}: memoryview non-contiguous")
+        if mv.ndim != 1:
+            raise ValueError(f"{i}: memoryview has {mv.ndim} dimensions, not 1")
+        if mv.format != format:
+            raise ValueError(f"{i}: inconsistent format: {mv.format} vs {format}")
+
+        start_addr = address_of_memoryview(mv)
+        if first_start_addr == 0:
+            first_start_addr = start_addr
+        else:
+            expected_addr = first_start_addr + nbytes
+            if start_addr != expected_addr:
+                raise ValueError(
+                    f"memoryview {i} does not start where the previous ends. "
+                    f"Expected {expected_addr:x}, starts {start_addr - expected_addr} byte(s) away."
+                )
+        nbytes += mv.nbytes
+
+    if nbytes == 0:
+        # all memoryviews were zero-length
+        assert len(first) == 0
+        return first
+
+    assert first_start_addr != 0, "Underlying buffer is null pointer?!"
+
+    base_mv = memoryview(obj).cast("B")
+    base_start_addr = address_of_memoryview(base_mv)
+    start_index = first_start_addr - base_start_addr
+
+    return base_mv[start_index : start_index + nbytes].cast(format)
+
+
+def address_of_memoryview(mv: memoryview) -> int:
+    """
+    Get the pointer to the first byte of a memoryview's data.
+
+    If the memoryview is read-only, NumPy must be installed.
+    """
+    # NOTE: this method relies on pointer arithmetic to figure out
+    # where each memoryview starts within the underlying buffer.
+    # There's no direct API to get the address of a memoryview,
+    # so we use a trick through ctypes and the buffer protocol:
+    # https://mattgwwalker.wordpress.com/2020/10/15/address-of-a-buffer-in-python/
+    one_byte_carr = ctypes.c_byte * 1
+    # ^ length and type don't matter, just use it to get the address of the first byte
+
+    try:
+        carr = one_byte_carr.from_buffer(mv)
+    except TypeError:
+        # `mv` is read-only. `from_buffer` requires the buffer to be writeable.
+        # See https://bugs.python.org/issue11427 for discussion.
+        # This typically comes from `deserialize_bytes`, where `mv.obj` is an
+        # immutable bytestring.
+        pass
+    else:
+        return ctypes.addressof(carr)
+
+    try:
+        import numpy as np
+    except ImportError:
+        raise ValueError(
+            f"Cannot get address of read-only memoryview {mv} since NumPy is not installed."
+        ) from None
```

Comment:
I figured the error message was self-explanatory and didn't need the ImportError chained onto it, why would you prefer to keep it?

Reply:
``Exception.__cause__`` is only set if you explicitly ask for it, e.g.

```python
try:
    ....
except Exception1 as e:
    raise Exception2(...) from e
```
Writing ``raise ... from None`` is redundant as it is the default.

Reply:
`__cause__` is None in either case, but when you explicitly set it to None, that also sets `__suppress_context__` to True, so the traceback is displayed as only the last exception. It's just a minor style thing; I think it's nice to shorten the displayed traceback when possible, so it's easier to understand if a user sees it. It's all irrelevant in this case anyway, since this error will always be swallowed in `merge_and_deserialize`.

```python
In [4]: try:
   ...:     import asdf
   ...: except ImportError:
   ...:     raise ValueError("couldn't import")
   ...: 
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-4-d39ad270c21f> in <module>
      1 try:
----> 2     import asdf
      3 except ImportError:

ModuleNotFoundError: No module named 'asdf'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-d39ad270c21f> in <module>
      2     import asdf
      3 except ImportError:
----> 4     raise ValueError("couldn't import")
      5 

ValueError: couldn't import

In [5]: try:
   ...:     import asdf
   ...: except ImportError:
   ...:     raise ValueError("couldn't import") from None
   ...: 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-28105f21527c> in <module>
      2     import asdf
      3 except ImportError:
----> 4     raise ValueError("couldn't import") from None
      5 

ValueError: couldn't import
```
```python
In [2]: try:
   ...:     try:
   ...:         import asdf
   ...:     except ImportError:
   ...:         raise ValueError("couldn't import")
   ...: except ValueError as e:
   ...:     print(f"{e.__cause__=}, {e.__context__=}, {e.__suppress_context__=}")
   ...: 
   ...: 
e.__cause__=None, e.__context__=ModuleNotFoundError("No module named 'asdf'"), e.__suppress_context__=False

In [3]: 

In [3]: try:
   ...:     try:
   ...:         import asdf
   ...:     except ImportError:
   ...:         raise ValueError("couldn't import") from None
   ...: except ValueError as e:
   ...:     print(f"{e.__cause__=}, {e.__context__=}, {e.__suppress_context__=}")
   ...: 
e.__cause__=None, e.__context__=ModuleNotFoundError("No module named 'asdf'"), e.__suppress_context__=True
```